vms/avm/service.go

```diff
@@ -231,6 +232,79 @@ func (service *Service) GetBalance(r *http.Request, args *GetBalanceArgs, reply
 	return nil
 }
 
+// Balance ...
+type Balance struct {
+	AssetID string      `json:"assetID"`
+	Balance json.Uint64 `json:"balance"`
+}
+
+// GetAllBalancesArgs are arguments for calling into GetAllBalances
+type GetAllBalancesArgs struct {
+	Address string `json:"address"`
+}
+
+// GetAllBalancesReply is the response from a call to GetAllBalances
+type GetAllBalancesReply struct {
+	Balances []Balance `json:"balances"`
+}
+
+// GetAllBalances returns a map where:
+//   Key: ID of an asset such that [args.Address] has a non-zero balance of the asset
+//   Value: The balance of the asset held by the address
+// Note that balances include assets that the address only _partially_ owns
+// (ie is one of several addresses specified in a multi-sig)
+func (service *Service) GetAllBalances(r *http.Request, args *GetAllBalancesArgs, reply *GetAllBalancesReply) error {
+	service.vm.ctx.Log.Verbo("GetAllBalances called with address: %s", args.Address)
+
+	address, err := service.vm.Parse(args.Address)
+	if err != nil {
+		return fmt.Errorf("couldn't parse given address: %s", err)
+	}
+	addrAsSet := ids.Set{}
+	addrAsSet.Add(ids.NewID(hashing.ComputeHash256Array(address)))
+
+	utxos, err := service.vm.GetUTXOs(addrAsSet)
+	if err != nil {
+		return fmt.Errorf("couldn't get address's UTXOs: %s", err)
+	}
+
+	assetIDs := ids.Set{}                    // IDs of assets the address has a non-zero balance of
+	balances := make(map[[32]byte]uint64, 0) // key: ID (as bytes). value: balance of that asset
+	for _, utxo := range utxos {
+		transferable, ok := utxo.Out.(ava.Transferable)
+		if !ok {
+			continue
+		}
+		assetID := utxo.AssetID()
+		assetIDs.Add(assetID)
+		balance := balances[assetID.Key()] // 0 if key doesn't exist
+		balance, err := safemath.Add64(transferable.Amount(), balance)
+		if err != nil {
+			balances[assetID.Key()] = math.MaxUint64
+		} else {
+			balances[assetID.Key()] = balance
+		}
+	}
+
+	avaAssetID, err := service.vm.Lookup("AVA")
+	if err != nil {
+		return errors.New("couldn't get asset ID of AVA")
+	}
+
+	reply.Balances = make([]Balance, assetIDs.Len())
+	for i, assetID := range assetIDs.List() {
+		var b Balance
+		if assetID.Equals(avaAssetID) {
+			b = Balance{AssetID: "AVA", Balance: json.Uint64(balances[assetID.Key()])}
+		} else {
+			b = Balance{AssetID: assetID.String(), Balance: json.Uint64(balances[assetID.Key()])}
+		}
+		reply.Balances[i] = b
+	}
```

Comment:
I think this should really be using `service.vm.PrimaryAlias(assetID)` rather than the special casing for "AVA".